### PR TITLE
Support colons in Middleware NextResponse.rewrite path

### LINKED ISSUE #31523: Middleware NextResponse.rewrite() crashes on colons in paths
### What version of Next.js are you using?

12.0.4

### What version of Node.js are you using?

16.10.0

### What browser are you using?

Chrome 95

### What operating system are you using?

macOS

### How are you deploying your application?

next start

### Describe the Bug

Using `NextResponse.rewrite("/:blahblah")` in a middleware to navigate to a path with a colon in it causes an obtuse `TypeError: Expected "blahblah" to be a string` style error(ultimately coming from `path-to-regexp`).

### Expected Behavior

Rewriting to a path with a colon in it to simply work without any unexpected replacement being attempted internally.

### To Reproduce

something like this in any middleware

```js
import { NextResponse } from "next/server";

export default function middleware() {
  return NextResponse.rewrite("/not-a:param");
}
```

repro repo here: https://github.com/klarstrup/next-middleware-rewrite-bug

### PR DESCRIPTION
**Pull request by klarstrup**
Fixes #31523 which was introduced here https://github.com/javivelasco/next.js/commit/8ae63b8c5ceb860116ea5d748e031d5ec054e86b#diff-ea09101c272788873ce720075e4839f0254179af0a07259a107752cafb193949R1230-R1235

It's problematic to be using code intended for the router with substitution handling etc. here, skipping `prepareDestination` is basically a soft revert of part of the above commit 

## Bug

- [x] Related issues linked using `fixes #number`
- [x] Integration tests added

## Documentation / Examples

- [x] Make sure the linting passes by running `yarn lint`

### CODE REVIEW

**Comment by klarstrup** [comment]
please advise @javivelasco

**Comment by pluma** [comment]
Is there any way to get official eyeballs on this PR? Seems like it addresses the colon problem.

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 8788b638ab84014093abf72a74ed2f7a01a15769

`yarn testheadless test/integration/amp-export-validation/test/index.test.js` 
- AMP Validation on Export > should have shown errors during build
- AMP Validation on Export > should export AMP pages

<details>
<summary>Expand output</summary>

  ● AMP Validation on Export › should have shown errors during build

    expect(received).toMatch(expected)

    Expected pattern: /error.*The mandatory attribute 'height' is missing in tag 'amp-video'\./
    Received string:  "info  - Checking validity of types...
    info  - Creating an optimized production build...
    info  - Compiled successfully
    info  - Collecting page data...
    info  - Generating static pages (0/8)
    info  - Generating static pages (2/8)
    info  - Generating static pages (4/8)
    info  - Generating static pages (6/8)
    info  - Generating static pages (8/8)
    Amp Validation·
    /cat          warn  The native JavaScript AMPHTML Validator (validator.js) has been turned down. If you are seeing this error, update your tooling to instead load the API compatible WebAssembly AMPHTML Validator (validator_wasm.js) instead.·
    /dog-cat      warn  The native JavaScript AMPHTML Validator (validator.js) has been turned down. If you are seeing this error, update your tooling to instead load the API compatible WebAssembly AMPHTML Validator (validator_wasm.js) instead.·
    /dog          warn  The native JavaScript AMPHTML Validator (validator.js) has been turned down. If you are seeing this error, update your tooling to instead load the API compatible WebAssembly AMPHTML Validator (validator_wasm.js) instead.·
    /first        warn  The native JavaScript AMPHTML Validator (validator.js) has been turned down. If you are seeing this error, update your tooling to instead load the API compatible WebAssembly AMPHTML Validator (validator_wasm.js) instead.·
    /second       warn  The native JavaScript AMPHTML Validator (validator.js) has been turned down. If you are seeing this error, update your tooling to instead load the API compatible WebAssembly AMPHTML Validator (validator_wasm.js) instead.·
    /third?amp=1  warn  The native JavaScript AMPHTML Validator (validator.js) has been turned down. If you are seeing this error, update your tooling to instead load the API compatible WebAssembly AMPHTML Validator (validator_wasm.js) instead.·
    info  - Finalizing page optimization...·
    Page                                       Size     First Load JS
    ┌ ○ /404                                   195 B          71.2 kB
    ├ ○ /cat (585 ms)                          AMP                AMP
    ├ ○ /dog                                   AMP                AMP
    ├ ○ /dog-cat                               AMP                AMP
    ├ ○ /first                                 AMP                AMP
    ├ ○ /second                                AMP                AMP
    └ ○ /third                                 280 B          71.3 kB
    + First Load JS shared by all              71.1 kB
      ├ chunks/framework-158f450427c55907.js   41.9 kB
      ├ chunks/main-080e92ea80d1caab.js        27 kB
      ├ chunks/pages/_app-26477814f4b13881.js  1.37 kB
      └ chunks/webpack-fd82975a6094609f.js     727 B·
    ○  (Static)  automatically rendered as static HTML (uses no initial props)·
    warn  - Detected next.config.js, no exported configuration found. https://nextjs.org/docs/messages/empty-configuration
    warn  - The Next.js plugin was not detected in your ESLint configuration. See https://nextjs.org/docs/basic-features/eslint#migrating-existing-config
    "

      23 |   })
      24 |
    > 25 |   it('should have shown errors during build', async () => {
         |                             ^
      26 |     expect(buildOutput).toMatch(
      27 |       /error.*The mandatory attribute 'height' is missing in tag 'amp-video'\./
      28 |     )

      at Object.<anonymous> (integration/amp-export-validation/test/index.test.js:25:29)

  ● AMP Validation on Export › should export AMP pages

    expect(received).toBe(expected) // Object.is equality

    Expected: "PASS"
    Received: {"i": "FAIL"}



      at Object.validateAMP (lib/amp-test-utils.js:26:27)
      at integration/amp-export-validation/test/index.test.js:35:13
          at async Promise.all (index 1)
      at Object.<anonymous> (integration/amp-export-validation/test/index.test.js:33:9)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| buildDuration | 15s | 15.2s | ⚠️ +209ms |
| buildDurationCached | 6s | 5.9s | -71ms |
| nodeModulesSize | 359 MB | 359 MB | ⚠️ +55 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.514 | 2.481 | -0.03 |
| / avg req/sec | 994.47 | 1007.48 | +13.01 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.163 | 1.161 | 0 |
| /error-in-render avg req/sec | 2148.89 | 2153.07 | +4.18 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 27.4 kB | 27.4 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.2 kB | 71.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 326 B | 326 B | ✓ |
| dynamic-HASH.js gzip | 2.37 kB | 2.37 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 919 B | 919 B | ✓ |
| image-HASH.js gzip | 4.94 kB | 4.94 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.19 kB | 2.19 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.3 kB | 14.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 546 B | 546 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| buildDuration | 18.4s | 18.2s | -167ms |
| buildDurationCached | 6s | 6s | -14ms |
| nodeModulesSize | 359 MB | 359 MB | ⚠️ +55 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.503 | 2.505 | 0 |
| / avg req/sec | 998.87 | 997.86 | ⚠️ -1.01 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.194 | 1.157 | -0.04 |
| /error-in-render avg req/sec | 2093.2 | 2161.13 | +67.93 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 27.5 kB | 27.5 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.4 kB | 71.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.36 kB | 2.36 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 911 B | 911 B | ✓ |
| image-HASH.js gzip | 4.98 kB | 4.98 kB | ✓ |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.21 kB | 2.21 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.3 kB | 14.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| index.html gzip | 530 B | 530 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 525 B | 525 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: 8788b638ab84014093abf72a74ed2f7a01a15769

**Review by javivelasco — COMMENTED** [review]
Gonna be checking this in detail asap

File: packages/next/server/next-server.ts [hunk 1]
```diff
@@ -1130,11 +1130,11 @@ export default class NextNodeServer extends BaseServer {
...
-          const { newUrl, parsedDestination } = prepareDestination({
-            appendParamsToQuery: false,
-            destination: rewritePath,
-            params: _params,
+          const parsedDestination = parseUrl(rewritePath)
+          const newUrl = formatUrl({
+            ...parsedDestination,
             query: {},
+            search: '',
           })
```
Comment (nkzawa) [inline packages/next/server/next-server.ts:1138]:
```suggestion
          const newUrl = parsedDestination.pathname
```

It looks like `newUrl` should contain only pathname?

**Review by nkzawa — CHANGES_REQUESTED** [review]
This looks awesome aside from a comment.

**Review by nkzawa — APPROVED** [review]

**Review by javivelasco — APPROVED** [review]
I think it looks good overall and all tests are passing so 🚀 
Thank you!

**Review by ijjk — APPROVED** [review]

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| buildDuration | 19.4s | 19.7s | ⚠️ +297ms |
| buildDurationCached | 7.6s | 7.7s | ⚠️ +34ms |
| nodeModulesSize | 372 MB | 372 MB | -198 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.878 | 3.838 | -0.04 |
| / avg req/sec | 644.68 | 651.34 | +6.66 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.679 | 1.686 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 1488.75 | 1482.75 | ⚠️ -6 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 27.9 kB | 27.9 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.5 kB | 71.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.57 kB | 2.57 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.06 kB | 5.06 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.26 kB | 2.26 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.8 kB | 14.8 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 525 B | 525 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| buildDuration | 23.1s | 23s | -101ms |
| buildDurationCached | 7.5s | 7.7s | ⚠️ +145ms |
| nodeModulesSize | 372 MB | 372 MB | -198 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.842 | 3.8 | -0.04 |
| / avg req/sec | 650.62 | 657.96 | +7.34 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.82 | 1.656 | -0.16 |
| /error-in-render avg req/sec | 1373.6 | 1509.8 | +136.2 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.1 kB | 28.1 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.1 kB | 72.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 324 B | 324 B | ✓ |
| dynamic-HASH.js gzip | 2.56 kB | 2.56 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.2 kB | 5.2 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.33 kB | 2.33 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 388 B | 388 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.9 kB | 14.9 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | klarstrup/next.js support-colons-in-rewrite-paths | Change |
| - | - | - | - |
| index.html gzip | 530 B | 530 B | ✓ |
| link.html gzip | 543 B | 543 B | ✓ |
| withRouter.html gzip | 525 B | 525 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: c3c6008cc97489c46b62e7388febf046130cd963

**Pull request merged**